ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParsePayload.java

```diff
@@ -208,7 +201,38 @@ private void addAttributesFromPayload(Map<String, String> attributes, ObjectNode
           .ifPresent(v -> attributes.put(Attribute.OS_VERSION, v));
     }
 
+    ParsePayload.addClientIdFromPayload(attributes, json);
+
+    // Add sample id, usually based on hashing clientId, but some other IDs are also supported to
+    // allow sampling on non-telemetry pings.
+    Stream.of(attributes.get(Attribute.CLIENT_ID),
+        // "impression_id" is a client_id-like identifier used in activity-stream ping
+        // that do not contain a client_id.
+        json.path("impression_id").textValue()) //
+        .map(ParsePayload::normalizeUuid) //
+        .filter(Objects::nonNull) //
+        .findFirst() //
+        .ifPresent(v -> attributes.put(Attribute.SAMPLE_ID, Long.toString(calculateSampleId(v))));
+  }
+
+  /** Extracts the client ID from the payload and adds it to `attributes`. */
+  public static void addClientIdFromPayload(Map<String, String> attributes, ObjectNode json) {
+    // Try to get glean-style client_info object.
+    JsonNode gleanClientInfo = json.path("client_info");
```

Comment:
let's also extract this one to a separate `getGleanClientInfo` function. It's short, but it's subject to the same hazard of introducing errors via repeated logic.